File: app/src/main/java/igrek/songbook/layout/songpreview/transpose/ChordsTransposerManager.java
package igrek.songbook.layout.songpreview.transpose;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.R;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.info.UiInfoService;
import igrek.songbook.info.UiResourceService;
import igrek.songbook.layout.songpreview.LyricsManager;
import igrek.songbook.layout.songpreview.SongPreviewLayoutController;
import igrek.songbook.layout.songpreview.view.quickmenu.QuickMenuTranspose;

public class ChordsTransposerManager {
	
	private int transposed = 0;
	private ChordsTransposer chordsTransposer = new ChordsTransposer();
	
	@Inject
	Lazy<LyricsManager> lyricsManager;
	@Inject
	Lazy<SongPreviewLayoutController> songPreviewController;
	@Inject
	UiResourceService uiResourceService;
	@Inject
	UiInfoService userInfo;
	@Inject
	Lazy<QuickMenuTranspose> quickMenuTranspose;
	
	public ChordsTransposerManager() {
		DaggerIoc.getFactoryComponent().inject(this);
	}
	
	public void reset() {
		transposed = 0;
	}
	
	public String transposeContent(String fileContent) {
		return chordsTransposer.transposeContent(fileContent, transposed);
	}
	
	public void onTransposeEvent(int semitones) {
		transposeBy(semitones);
		
		songPreviewController.get().onCrdModelUpdated();
		
		String info = uiResourceService.resString(R.string.transposed_by_semitones, getTransposedByDisplayName());
		
		if (getTransposedBy() != 0) {
			userInfo.showInfoWithAction(info, R.string.action_transposition_reset, this::onTransposeResetEvent);
		} else {
			userInfo.showInfo(info);
		}
		
		quickMenuTranspose.get().onTransposedEvent();
	}
	
	public void onTransposeResetEvent() {
		onTransposeEvent(-getTransposedBy());
	}
	
	private void transposeBy(int semitones) {
		transposed += semitones;
		if (transposed >= 12)
			transposed -= 12;
		if (transposed <= -12)
			transposed += 12;
		lyricsManager.get().reparse();
	}
	
	private int getTransposedBy() {
		return transposed;
	}
	
	public String getTransposedByDisplayName() {
		return (transposed > 0 ? "+" : "") + transposed + " " + getSemitonesDisplayName(transposed);
	}
	
	private String getSemitonesDisplayName(int transposed) {
		int absTransposed = transposed >= 0 ? transposed : -transposed;
		int stringResId;
		if (absTransposed == 0) {
			stringResId = R.string.transpose_0_semitones;
		} else if (absTransposed == 1) {
			stringResId = R.string.transpose_1_semitones;
		} else if (absTransposed <= 4) { // 2,3,4
			stringResId = R.string.transpose_234_semitones;
		} else {
			stringResId = R.string.transpose_5_semitones;
		}
		return uiResourceService.resString(stringResId);
	}
	
}

File: app/src/main/java/igrek/songbook/chordsv2/loader/ChordsTransposerManager.kt
package igrek.songbook.chordsv2.loader

import igrek.songbook.R
import igrek.songbook.info.UiInfoService
import igrek.songbook.info.UiResourceService
import igrek.songbook.inject.LazyExtractor
import igrek.songbook.inject.LazyInject
import igrek.songbook.inject.appFactory
import igrek.songbook.persistence.repository.SongsRepository
import igrek.songbook.songpreview.SongPreviewLayoutController
import igrek.songbook.songpreview.quickmenu.QuickMenuTranspose

class ChordsTransposerManager(
    lyricsLoader: LazyInject<LyricsLoader> = appFactory.lyricsLoader,
    songPreviewLayoutController: LazyInject<SongPreviewLayoutController> = appFactory.songPreviewLayoutController,
    uiResourceService: LazyInject<UiResourceService> = appFactory.uiResourceService,
    uiInfoService: LazyInject<UiInfoService> = appFactory.uiInfoService,
    quickMenuTranspose: LazyInject<QuickMenuTranspose> = appFactory.quickMenuTranspose,
    songsRepository: LazyInject<SongsRepository> = appFactory.songsRepository,
) {
    private val lyricsLoader by LazyExtractor(lyricsLoader)
    private val songPreviewController by LazyExtractor(songPreviewLayoutController)
    private val uiResourceService by LazyExtractor(uiResourceService)
    private val userInfo by LazyExtractor(uiInfoService)
    private val quickMenuTranspose by LazyExtractor(quickMenuTranspose)
    private val songsRepository by LazyExtractor(songsRepository)

    var transposedBy = 0
    val isTransposed: Boolean = transposedBy != 0

    val transposedByDisplayName: String = when {
        transposedBy > 0 ->"+$transposedBy ${getSemitonesDisplayName(transposedBy)}"
        else -> "$transposedBy ${getSemitonesDisplayName(transposedBy)}"
    }

    fun reset(initialTransposed: Int = 0) {
        transposedBy = initialTransposed
    }

    fun onTransposeEvent(semitones: Int) {
        transposeBy(semitones)

        songPreviewController.onLyricsModelUpdated()

        if (isTransposed) {
            userInfo.showInfoAction(
                R.string.transposed_by_semitones, transposedByDisplayName,
                actionResId = R.string.action_transposition_reset
            ) {
                this.onTransposeResetEvent()
            }
        } else {
            userInfo.showInfo(R.string.transposed_by_semitones, transposedByDisplayName)
        }

        quickMenuTranspose.onTransposedEvent()

        val song = songPreviewController.currentSong
        if (song != null)
            songsRepository.transposeDao.setSongTransposition(song.songIdentifier(), transposedBy)
    }

    fun onTransposeResetEvent() {
        transposeBy(-transposedBy)
        onTransposeEvent(-transposedBy)
    }

    private fun transposeBy(semitones: Int) {
        transposedBy += semitones
        if (transposedBy >= 12)
            transposedBy -= 12
        if (transposedBy <= -12)
            transposedBy += 12

        lyricsLoader.onTransposed()
    }

    private fun getSemitonesDisplayName(transposed: Int): String {
        val absTransposed = if (transposed >= 0) transposed else -transposed
        val stringResId: Int = when {
            absTransposed == 0 -> R.string.transpose_0_semitones
            absTransposed == 1 -> R.string.transpose_1_semitones
            absTransposed <= 4 -> // 2,3,4
                R.string.transpose_234_semitones
            else -> R.string.transpose_5_semitones
        }
        return uiResourceService.resString(stringResId)
    }

}